op-node/rollup/derive/channel.go

```diff
@@ -58,26 +53,47 @@ func (ch *Channel) AddFrame(frame Frame, l1InclusionBlock eth.L1BlockRef) error
 	if frame.ID != ch.id {
 		return fmt.Errorf("frame id does not match channel id. Expected %v, got %v", ch.id, frame.ID)
 	}
+	// These checks are specified and cannot be changed without a hard fork.
 	if frame.IsLast && ch.closed {
 		return fmt.Errorf("cannot add ending frame to a closed channel. id %v", ch.id)
 	}
 	if _, ok := ch.inputs[uint64(frame.FrameNumber)]; ok {
 		return DuplicateErr
 	}
-	// TODO: highest seen frame vs endFrameNumber
+	if ch.closed && frame.FrameNumber >= ch.endFrameNumber {
+		return fmt.Errorf("frame number (%d) is greater than or equal to end frame number (%d) of a closed channel", frame.FrameNumber, ch.endFrameNumber)
+	}
 
 	// Guaranteed to succeed. Now update internal state
 	if frame.IsLast {
 		ch.endFrameNumber = frame.FrameNumber
 		ch.closed = true
 	}
+	// Prune frames with a number higher than the closing frame number when we receive a closing frame
+	if frame.IsLast && ch.endFrameNumber < ch.highestFrameNumber {
+		// Do a linear scan over saved inputs instead of ranging over ID numbers
+		var idsToPrune []uint64
+		for id, prunedFrame := range ch.inputs {
+			if id >= uint64(ch.endFrameNumber) {
+				idsToPrune = append(idsToPrune, id)
+			}
+			ch.size -= frameSize(prunedFrame)
+		}
+		for _, id := range idsToPrune {
+			delete(ch.inputs, id)
+		}
+		ch.highestFrameNumber = ch.endFrameNumber
```

Comment:
In Go it's save to delete from a map while iterating the map, so we can avoid allocating idsToPrune here I think.

```suggestion
		// Do a linear scan over saved inputs instead of ranging over ID numbers
		for id, prunedFrame := range ch.inputs {
			if id >= uint64(ch.endFrameNumber) {
				delete(ch.inputs, id)
			}
			ch.size -= frameSize(prunedFrame)
		}
		ch.highestFrameNumber = ch.endFrameNumber
```